### PR TITLE
Correctly handle UI hierarchy without a camera

### LINKED ISSUE #12184: Panic with UI hierarchy when no camera is present
## Bevy version

main, 0.13

bisected to #10559

## Relevant system information

```
AdapterInfo { name: "Apple M1 Max", vendor: 0, device: 0, device_type: IntegratedGpu, driver: "", driver_info: "", backend: Metal }
SystemInfo { os: "MacOS 14.2.1 ", kernel: "23.2.0", cpu: "Apple M1 Max", core_count: "10", memory: "64.0 GiB" }
```

## What you did

I noticed this while writing some doc examples which were accidentally not `no_run`.

This is the most minimal repro I can make, but I also see this panic in e.g. the `button` example modified with no camera.

```rust
use bevy::prelude::*;

fn main() {
    App::new()
        .add_plugins(DefaultPlugins)
        .add_systems(Startup, setup)
        .run();
}

fn setup(mut commands: Commands) {
    commands
        .spawn(NodeBundle::default())
        .with_children(|parent| {
            parent.spawn(NodeBundle::default());
        });
}
```

## What went wrong

Received a warning about a camera not being present, which I suppose makes sense.

Received a warning about an "unstyled child in a UI entity hierarchy" which does not make sense as far as I can tell.

Finally, panicked on an `unwrap` which doesn't seem desirable.

```
2024-02-28T15:05:02.776557Z  WARN bevy_ui::layout: No camera found to render UI to. To fix this, add at least one camera to the scene.
2024-02-28T15:05:02.776596Z  WARN bevy_ui::layout: Unstyled child in a UI entity hierarchy. You are using an entity without UI components as a child of an entity with UI components, results may be unexpected.
thread 'Compute Task Pool (0)' panicked at crates/bevy_ui/src/layout/mod.rs:131:60:
called `Option::unwrap()` on a `None` value
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Encountered a panic in system `bevy_ui::layout::ui_layout_system`!
Encountered a panic in system `bevy_app::main_schedule::Main::run_main`!
thread 'main' panicked at /Users/me/.cargo/registry/src/index.crates.io-6f17d22bba15001f/winit-0.29.11/src/platform_impl/macos/app_state.rs:387:33:
called `Result::unwrap()` on an `Err` value: PoisonError { .. }
```

### PR DESCRIPTION
**Pull request by salvadorcarvalhinho**
# Objective
Add support so bevy_ui can correctly handle an UI hierarchy without a camera present.

- Fixes #12184

## Solution

As there was no default behavior for what should happen when a camera is not present in a UI hierarchy, the solution 
was based in defining that default behavior and improving the overall handling of this "exception".

## Changelog

- Create default values to be used in upsert_node
- Add flag to control warnings about no camera present
- Create unit test no_camera_ui (to test if ui handles no camera present)

### CODE REVIEW

File: crates/bevy_ui/src/layout/mod.rs [hunk 1]
```diff
@@ -330,14 +340,18 @@ pub fn ui_layout_system(
...
                 let layout_info = camera_layout_info
                     .entry(camera_entity)
                     .or_insert_with(|| calculate_camera_layout_info(camera));
                 layout_info.root_nodes.push(entity);
             }
             None => {
                 if cameras.is_empty() {
-                    warn!("No camera found to render UI to. To fix this, add at least one camera to the scene.");
+                    if !ui_surface.no_camera {
+                        ui_surface.no_camera = true;
```
Comment (StrikeForceZero) [inline crates/bevy_ui/src/layout/mod.rs:352]: IMO the `no_camera` flag is not needed and just adds an extra layer. I tested it with this diff below applied to your branch and it still passed the test. But the else upsert_node below with the default style and layout is a good catch. :+1:

If we are worried about spamming the console we could use a `warn_once!` instead.

```dif
Index: crates/bevy_ui/src/layout/mod.rs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/crates/bevy_ui/src/layout/mod.rs b/crates/bevy_ui/src/layout/mod.rs
--- a/crates/bevy_ui/src/layout/mod.rs	(revision 8bbf91532c1fcdef9949ae021f3c3531c970c865)
+++ b/crates/bevy_ui/src/layout/mod.rs	(date 1711922950048)
@@ -63,7 +63,6 @@
     camera_entity_to_taffy: EntityHashMap<EntityHashMap<taffy::node::Node>>,
     camera_roots: EntityHashMap<Vec<RootNodePair>>,
     taffy: Taffy,
-    no_camera: bool,
 }
 
 fn _assert_send_sync_ui_surface_impl_safe() {
@@ -91,7 +90,6 @@
             camera_entity_to_taffy: Default::default(),
             camera_roots: Default::default(),
             taffy,
-            no_camera: false,
         }
     }
 }
@@ -340,7 +338,6 @@
                     );
                     continue;
                 };
-                ui_surface.no_camera = false;
                 let layout_info = camera_layout_info
                     .entry(camera_entity)
                     .or_insert_with(|| calculate_camera_layout_info(camera));
@@ -348,10 +345,7 @@
             }
             None => {
                 if cameras.is_empty() {
-                    if !ui_surface.no_camera {
-                        ui_surface.no_camera = true;
-                        warn!("No camera found to render UI to. To fix this, add at least one camera to the scene.");
-                    }
+                    warn!("No camera found to render UI to. To fix this, add at least one camera to the scene.");
                 } else {
                     warn!(
                         "Multiple cameras found, causing UI target ambiguity. \
```

Comment (salvadorcarvalhinho) [reply crates/bevy_ui/src/layout/mod.rs:352]: That makes perfect sense, I'll make the commit with that changes! Thank you for your help!

File: crates/bevy_ui/src/layout/mod.rs [hunk 2]
```diff
@@ -30,6 +30,14 @@ pub struct LayoutContext {
...
 
 impl LayoutContext {
+    fn default() -> Self {
+        Self {
+            scale_factor: 1.0,
+            physical_size: Vec2::ZERO,
+            min_size: 0.0,
+            max_size: 0.0,
+        }
+    }
```
Comment (james7132) [inline crates/bevy_ui/src/layout/mod.rs:40]: This should be a `Default` trait impl.

Comment (salvadorcarvalhinho) [reply crates/bevy_ui/src/layout/mod.rs:40]: Yes, that make sense! Added those changes in the newest commit!

Comment (james7132) [reply crates/bevy_ui/src/layout/mod.rs:40]: No, as in `LayoutContext` should implement Default instead of a `default` method like this.

Comment (StrikeForceZero) [reply crates/bevy_ui/src/layout/mod.rs:40]:
```rust
impl Default for LayoutContext {
   fn default() -> Self {
        Self {
            scale_factor: 1.0,
            physical_size: Vec2::ZERO,
            min_size: 0.0,
            max_size: 0.0,
        }
   }
}
```

Comment (salvadorcarvalhinho) [reply crates/bevy_ui/src/layout/mod.rs]: Oh, I misunderstood your sugestion, thanks for clarifying! 
The last commit should already have those changes implemented.

**Review by james7132 — APPROVED** [review]

**Review by StrikeForceZero — APPROVED** [review]
The merge of the `main` into the PR's branch makes me nervous but I think all PR's get squashed so it shouldn't matter. Everything else looks good.

**Review by alice-i-cecile — APPROVED** [review]

**Pull request merged**